src/user.rs

```diff
@@ -0,0 +1,177 @@
+
+use validator::Validate;
+use validator_derive;
+
+use reqwest;
+use serde::{Deserialize, Serialize};
+
+/// Submits a GET request to a specified URL, and returns a vector of users. Expected response is in json format, and can be parsed by the `User` struct with accessible elements by key name.
+pub async fn get_users(url: &str) -> Result<Vec<User>, Box<dyn std::error::Error>> {
+    let users = reqwest::get(url).await?.json::<Vec<User>>().await?;
+
+    Ok(users)
+}
+
+/// `POST` a new user with a JSON formatted raw string (&str) passed into `post_user`. Returns a `User` struct with fields filled as returned by the endpoint,if the post is valid. This will include a new `User.id` field if the user is set properly.
+///
+/// The name will need to be at least 3 characters long and the email address should be valid.
+///
+/// ## Errors
+///
+/// This will panic if the `name` is less than 3 characters, or `email` is not valid.
+///
+/// > Note: The name will need to be at least 3 characters long and the email address should be valid.
+pub async fn post_user(url: &str, user_json: &str) -> Result<User, Box<dyn std::error::Error>> {
+    let post: User = serde_json::from_str(user_json)?;
+
+    post.validate()?; // validation requirements defined in struct definition
+
+    let client = reqwest::Client::new();
+    let resp = client.post(url).json(&post).send().await?;
+
+    let body = serde_json::from_str(resp.text().await?.as_str())?;
+
+    Ok(body)
+}
+
+/// Defining Structs used to serialize and deserialize responses from reqwest for a user
+#[derive(Serialize, Deserialize, validator_derive::Validate, Default, Debug)]
+pub struct User {
+    #[serde(default)]
+    pub id: usize,
+    #[serde(default)]
+    #[validate(length(min = 3))]
+    pub name: String,
+    #[serde(default)]
+    pub username: String,
+    #[serde(default)]
+    #[validate(email)]
+    pub email: String,
+    #[serde(default)]
+    pub address: Address,
+    #[serde(default)]
+    pub phone: String,
+    #[serde(default)]
+    pub website: String,
+    #[serde(default)]
+    pub company: Company,
+}
+
+#[derive(Serialize, Deserialize, Default, Debug)]
+pub struct Address {
+    #[serde(default)]
+    pub street: String,
+    #[serde(default)]
+    pub suite: String,
+    #[serde(default)]
+    pub city: String,
+    #[serde(default)]
+    pub zipcode: String,
+    #[serde(default)]
+    pub geo: Geo,
+}
+
+#[derive(Serialize, Deserialize, Default, Debug)]
+pub struct Geo {
+    #[serde(default)]
+    pub lat: String,
+    #[serde(default)]
+    pub lng: String,
+}
+
+#[derive(Serialize, Deserialize, Default, Debug)]
+pub struct Company {
+    #[serde(default)]
+    pub name: String,
+    #[serde(rename = "catchPhrase", default)]
+    //snake case to camel for JSON format to keep idiomatic rust var names.
+    pub catch_phrase: String, // NOTE: JSON uses catchPhrase
+    #[serde(default)]
+    pub bs: String,
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Cool, these are good.  Definitely remove the code in main.rs that gets and posts.